server/src-lib/Hasura/Server/Query.hs

```diff
@@ -369,7 +384,18 @@ runQueryM rq =
 
       RQRunSql q                   -> runRunSQL q
 
-      RQBulk qs                    -> encJFromList <$> indexedMapM runQueryM qs
+      RQBulk qs                    -> assertAllTxAccess (zip [0::Integer ..] qs) >>
+                                      encJFromList <$> indexedMapM runQueryM qs
+      where
+        assertAllTxAccess = \case
+          [] -> pure ()
+          (_q:[]) -> pure ()
+          (q1:q2:qs) -> assertSameTxAccess q1 q2 >> assertAllTxAccess (q2:qs)
+
+        assertSameTxAccess (i1, q1) (i2, q2) = unless (getQueryAccessMode q1 == getQueryAccessMode q2) $
```

Comment:
This could be altered to be more of a “parser” than a “validator” (using the terminology from [Parse, Don’t Validate](https://lexi-lambda.github.io/blog/2019/11/05/parse-don-t-validate/)) by combining it with `getQueryAccessMode` and giving that function this type:

```haskell
getQueryAccessMode :: (MonadError QErr m) => RQLQuery -> m Q.TxAccess
```

That way, it’s impossible (or at least much harder) to accidentally omit the consistency check.